src/lib/tls/tls_server.cpp

```diff
@@ -1020,6 +1064,126 @@ void Server::session_create(Server_Handshake_State& pending_state,
 
    pending_state.server_hello_done(new Server_Hello_Done(pending_state.handshake_io(), pending_state.hash()));
    }
+
+DTLS_Prestate Server::pre_verify_cookie(Credentials_Manager& creds,
+                               const Policy& policy,
+                               const std::string& client_identity,
+                               const uint8_t input[],
+                               size_t buf_size,
+                               output_fn emit_data_fn)
+   {
+   secure_vector<uint8_t> readbuf;
+   secure_vector<uint8_t> record_buf;
+   Stream_Sequence_Numbers seq;
+   // a hacky use of Stream_Sqeuence_Numbers
+   SymmetricKey cookie_secret;
+   size_t consumed = 0;
+
+   const Record_Header record =
+      read_record(true, readbuf, input,
+                  buf_size, consumed,
+                  record_buf, &seq, [](uint16_t) -> std::shared_ptr<Connection_Cipher_State> {return nullptr;},
+                  true);
+   // should read_record be refactored to get rid of those quirks?
+
+   if (record.type() != HANDSHAKE)
+      {
+      throw TLS_Exception(Alert::Type::UNEXPECTED_MESSAGE,
+         "Not a handshake record");
+      }
+
+   if (record_buf.size() < 12)
+      {
+      throw TLS_Exception(Alert::Type::DECODE_ERROR,
+         "Message is too short");
+      }
+
+   const Handshake_Type msg_type =
+      static_cast<Handshake_Type>(record_buf[0]);
+
+   if (msg_type != CLIENT_HELLO)
+      {
+      throw TLS_Exception(Alert::Type::UNEXPECTED_MESSAGE,
+         "Not a client hello");
+      }
+
+   const size_t length = 4 + make_uint32(0, record_buf[1], record_buf[2], record_buf[3]);
+
+   if (length <= 4 || record_buf.size() != length + 8)
+      {
+      throw TLS_Exception(Alert::Type::DECODE_ERROR,
+         "Bad length in client hello");
+      }
+
+   const uint16_t msg_seq = make_uint16(record_buf[4], record_buf[5]);
+
+   if (msg_seq > 2)
+      {
+      // is it possible for a Client Hello to get beyond seq=2?
```

Comment:
I think yes if retransmitted/lost, eg RFC 6347 section 4.2.2

>   Note, however, that from the perspective of the DTLS record layer,
   the retransmission is a new record.  This record will have a new
   DTLSPlaintext.sequence_number value.

So it could go all the way to 2**48-1. [I mean not really, but at a protocol layer it is possible]

Also should verify here that epoch == 0.

Reply:
We are reading the sequence in the message here, not the record. And yes it is possible to get beyond 2 - when the server sends more than one Verify Request here, and I wonder if `Datagram_Sequence_Numbers::read_accept` calls in `Channel::create_handshake_state` should be made O(1) to avoid DoS possibility.

Reply:
Oh right sorry. But as you point out we can ping-pong repeatedly if eg cookie is corrupted several times, cookie key is rotated, etc, and so end up with a higher message sequence.

Good point on read_accept complexity - a client could send a (valid) client hello with a very high record sequence number and then we would spin.